src/types/middleware.ts

```diff
@@ -0,0 +1,28 @@
+import { EffectScope } from 'vue'
+import { MaybePromise } from '@/types/utilities'
+
+type Route = {
+  name: string,
+  params: Record<string, unknown>,
+  query: unknown,
+  hash: string,
+}
```

Comment:
We need to define this better. We'll want some type for the route that has been "resolved". I think the RouteFlat type might be what is used here but I'm not sure.

Reply:
I designed some components at my previous company that had types that built off a base and was really nice to work with, maybe we can do something similar here?

in this case I think the base is something like

```ts
type Route = {
  name: string,
  ...
}
```

and then we build on that with something like

```ts
type Resolved<T extends Route> = T & {
  params: Record<string, Param[]>
  ...
}
```

so components/services that expect a resolved route would be written as

```ts
function middleware(route: Resolved<Route>): void
```

I could see this being too "inheritance" for your liking, just throwing it out there as a possible syntax

Reply:
No I like that, I wanna see what that looks like.